### ISSUE TITLE
Crash 2.0.7 / Android / image / "Make sure you've a file explorer installed"

### ISSUE THREAD

**Issue by monsieurtanuki** [miguelpruivo/flutter_file_picker#440]
**Describe the bug**
Immediate crash when running this code on android Simulator (crashed too on smartphone):
```dart
FilePickerResult result = await FilePicker.platform.pickFiles(type: FileType.image);
```

```
D/FilePickerDelegate(12612): Selected type image/*
E/FilePickerDelegate(12612): Can't find a valid activity to handle the request. Make sure you've a file explorer installed.
I/flutter (12612): [MethodChannelFilePicker] Platform exception: PlatformException(invalid_format_type, Can't handle the provided file type., null, null)
E/flutter (12612): [ERROR:flutter/lib/ui/ui_dart_state.cc(177)] Unhandled Exception: PlatformException(invalid_format_type, Can't handle the provided file type., null, null)
```

**Error Log**
<details>

````
E/flutter (12612): #0      StandardMethodCodec.decodeEnvelope (package:flutter/src/services/message_codecs.dart:582:7)
E/flutter (12612): #1      MethodChannel._invokeMethod (package:flutter/src/services/platform_channel.dart:159:18)
E/flutter (12612): <asynchronous suspension>
E/flutter (12612): #2      MethodChannel.invokeMethod (package:flutter/src/services/platform_channel.dart:332:12)
E/flutter (12612): #3      MethodChannel.invokeListMethod (package:flutter/src/services/platform_channel.dart:345:41)
E/flutter (12612): #4      FilePickerIO._getPath (package:file_picker/src/file_picker_io.dart:79:47)
E/flutter (12612): #5      FilePickerIO.pickFiles (package:file_picker/src/file_picker_io.dart:29:7)
E/flutter (12612): #6      CurrentPhoto.openFilePickerAndSave (package:sellos/models/current_photo.dart:48:63)
E/flutter (12612): #7      NewPictureMenuButton.build.<anonymous closure> (package:sellos/sheets/new_picture_menu_button.dart:15:32)
E/flutter (12612): #8      _InkResponseState._handleTap (package:flutter/src/material/ink_well.dart:993:19)
E/flutter (12612): #9      _InkResponseState.build.<anonymous closure> (package:flutter/src/material/ink_well.dart:1111:38)
E/flutter (12612): #10     GestureRecognizer.invokeCallback (package:flutter/src/gestures/recognizer.dart:183:24)
E/flutter (12612): #11     TapGestureRecognizer.handleTapUp (package:flutter/src/gestures/tap.dart:598:11)
E/flutter (12612): #12     BaseTapGestureRecognizer._checkUp (package:flutter/src/gestures/tap.dart:287:5)
E/flutter (12612): #13     BaseTapGestureRecognizer.handlePrimaryPointer (package:flutter/src/gestures/tap.dart:222:7)
E/flutter (12612): #14     PrimaryPointerGestureRecognizer.handleEvent (package:flutter/src/gestures/recognizer.dart:476:9)
E/flutter (12612): #15     PointerRouter._dispatch (package:flutter/src/gestures/pointer_router.dart:77:12)
E/flutter (12612): #16     PointerRouter._dispatchEventToRoutes.<anonymous closure> (package:flutter/src/gestures/pointer_router.dart:122:9)
E/flutter (12612): #17     _LinkedHashMapMixin.forEach (dart:collection-patch/compact_hash.dart:377:8)
E/flutter (12612): #18     PointerRouter._dispatchEventToRoutes (package:flutter/src/gestures/pointer_router.dart:120:18)
E/flutter (12612): #19     PointerRouter.route (package:flutter/src/gestures/pointer_router.dart:106:7)
E/flutter (12612): #20     GestureBinding.handleEvent (package:flutter/src/gestures/binding.dart:358:19)
E/flutter (12612): #21     GestureBinding.dispatchEvent (package:flutter/src/gestures/binding.dart:338:22)
E/flutter (12612): #22     RendererBinding.dispatchEvent (package:flutter/src/rendering/binding.dart:267:11)
E/flutter (12612): #23     GestureBinding._handlePointerEvent (package:flutter/src/gestures/binding.dart:295:7)
E/flutter (12612): #24     GestureBinding._flushPointerEventQueue (package:flutter/src/gestures/binding.dart:240:7)
E/flutter (12612): #25     GestureBinding._handlePointerDataPacket (package:flutter/src/gestures/binding.dart:213:7)
E/flutter (12612): #26     _rootRunUnary (dart:async/zone.dart:1206:13)
E/flutter (12612): #27     _CustomZone.runUnary (dart:async/zone.dart:1100:19)
E/flutter (12612): #28     _CustomZone.runUnaryGuarded (dart:async/zone.dart:1005:7)
E/flutter (12612): #29     _invoke1 (dart:ui/hooks.dart:265:10)
E/flutter (12612): #30     _dispatchPointerDataPacket (dart:ui/hooks.dart:174:5)
````
</details>

**Flutter Version details**
<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel stable, 1.22.1, on Mac OS X 10.15.5 19F101, locale fr-FR)
    • Flutter version 1.22.1 at /Users/fabricefontaine/development/flutter
    • Framework revision f30b7f4db9 (3 days ago), 2020-10-08 10:06:30 -0700
    • Engine revision 75bef9f6c8
    • Dart version 2.10.1

 
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/fabricefontaine/Library/Android/sdk
    • Platform android-30, build-tools 30.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

 
[✓] Xcode - develop for iOS and macOS (Xcode 12.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.0.1, Build version 12A7300
    • CocoaPods version 1.9.3

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 50.0.1
    • Dart plugin version 193.7547
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

 
[✓] Connected device (1 available)            
    • sdk gphone x86 (mobile) • emulator-5554 • android-x86 • Android 11 (API 30) (emulator)
```
</details>

**Comment by miguelpruivo** [comment]
Hi @monsieurtanuki, the error is pretty clear here. You must have a valid file explorer to handle the picking for you. The plugin only asks the OS for a file explorer application, which is good because lets the users use whatever they want (or even their own explorers).

Most android devices have one by default, so you’re likely using a device that don’t. You can install for example ES FileExplorer which is free and it should work as intended.

Closing as this represents a non-issue.

Thank you!

**Comment by monsieurtanuki** [comment]
Thanks you @miguelpruivo. For the record:
* it crashed on my smartphone too and I assumed it was for the same reason, but when I looked in details the "smartphone crash" was my mistake, nothing 100% related
* there's no way I would ask an end-user to install a file explorer - though I understand that in some cases people may have a favorite file explorer (to be called/selected via an Android Intent I guess)
* if I use flutter library *image_picker* it works (on the same android emulator where *file_picker* crashes in this issue), opening the following file explorer
![Screenshot_1602433450](https://user-images.githubusercontent.com/11576431/95684029-03933380-0bef-11eb-84ad-b07d50cae962.png)

**Comment by miguelpruivo** [comment]
@monsieurtanuki that’s because image_picker uses an intent that opens images exclusively.

If you use file_picker with `pickFiles(type: FileType.image`) you’ll get the same result. On android there’s no “magic” involved. Either the system has a way to handle the intent or not. The other way around would be if the plugin had its own picker implemented, but that would had other downsides, such as not having the feeling & look that the dev might want or could potentially introduce some bug/unsupported feature.

As for the image_picker, it doesn’t implement its own picker, it behaves strictly like a file_picker exclusively for images/videos.

**Comment by monsieurtanuki** [comment]
@miguelpruivo As I said before I did use `pickFiles(type: FileType.image)` in *file_picker*, but that crashed.

**Comment by miguelpruivo** [comment]
@monsieurtanuki the screenshot you’ve posted is what should be open as well with that type. Try setting allowMultiple to see if it makes any difference.

One thing that could have impact is if image_picker is using ACTION_OPEN_DOCUMENT instead of ACTION_GET_CONTENT (which file_picker uses). Both have slightly differences for the same end. However, it would surprise me if the later is the cause of a crash for you since it’s been like that since ever.

Might be related with SDK 30. Could you please try with SDK 29 and see if the same happens to dig down the issue?

Thank you.

**Comment by monsieurtanuki** [comment]
@miguelpruivo I wouldn't know how to test again with SDK 29.
During my tests I tried with both 29 then 30 as `targetSdkVersion` and `compileSdkVersion` (same negative results).
The android simulator I used was Pixel 2: I guess you should be able to reproduce the issue with a fresh new simulator.

Regarding *image_picker* they seem to use `ACTION_GET_CONTENT` too (which makes sense): https://github.com/flutter/plugins/blob/master/packages/image_picker/image_picker/android/src/main/java/io/flutter/plugins/imagepicker/ImagePickerDelegate.java#L334
But they don't seem to try to resolve the activity/intent in that specific case, perhaps assuming that it's always supposed to work.

**Comment by miguelpruivo** [comment]
@monsieurtanuki creating a simulator with SDK 29 should be enough but I’ll give it a shot as soon as I can.

Thank you for the input.

**Comment by delaosa** [comment]
Same issue here, it's working well with API 29 however with API 30 I'm getting same error than @monsieurtanuki.

**Comment by miguelpruivo** [comment]
@delaosa thank you for the input. Might be related with sdk 30. Should be an easy fix. I’ll take a look ASAP.

**Comment by miguelpruivo** [comment]
![Captura de ecrã 2020-10-19, às 20 57 42](https://user-images.githubusercontent.com/27860743/96505385-f52bc400-124d-11eb-8569-dbc5c46b95ad.png)
![Captura de ecrã 2020-10-19, às 20 57 18](https://user-images.githubusercontent.com/27860743/96505389-f5c45a80-124d-11eb-825f-3a65200af347.png)
@monsieurtanuki using a simulator with sdk 30 and using `FileType.image` results in the expected behavior for me (just like you have on `image_picker`) check the screenshots above. 

I thought that at first could be something related to it, but turns out that all seem to be working as expected. What could it be?

**Comment by monsieurtanuki** [comment]
@miguelpruivo I guess that your emulator has at least one app that matches the intent, and mine has not (officially).
Which explains why it works for you, it doesn't work for me, and image_picker works for me (they don't test at all but the action works all the same).

Anyway I think you should NOT crash even if the test is false, you should return gracefully instead of raising an Exception (if I remember well).

In my case for some reason the test is false (no app matches the intent) but the action does work (image selection).
The test is apparently both irrelevant and dangerous on my emulator: how relevant is that test for all android devices, more generally speaking? What does internet say about that?

I will indeed test only your sample code, but I would be very surprised if I get different results.

**Comment by miguelpruivo** [comment]
@monsieurtanuki rule thumb is that if I’m not able to replicate it (even when the scenarios are identical) makes it very difficult to identify and fix the issue, hence every bit counts.

I had a fresh Pixel 2 Android 11 emulator as of today’s test (didn’t delete or installed anything). Did you delete?

**Comment by monsieurtanuki** [comment]
@miguelpruivo It looks similar to:
* https://stackoverflow.com/questions/62535856/intent-resolveactivity-returns-null-in-api-30
* https://developer.android.com/about/versions/11/privacy/package-visibility

If needed:
<details>
<summary>my emulator details</summary>

```
Name: Pixel_2_API_30
CPU/ABI: Google Play Intel Atom (x86)
Path: /Users/fabricefontaine/.android/avd/Pixel_2_API_30.avd
Target: google_apis_playstore [Google Play] (API level 30)
Skin: pixel_2
SD Card: 512M
fastboot.chosenSnapshotFile: 
runtime.network.speed: full
hw.accelerometer: yes
hw.device.name: pixel_2
hw.lcd.width: 1080
hw.initialOrientation: Portrait
image.androidVersion.api: 30
tag.id: google_apis_playstore
hw.mainKeys: no
hw.camera.front: emulated
avd.ini.displayname: Pixel 2 API 30
hw.gpu.mode: auto
hw.ramSize: 1536
PlayStore.enabled: true
fastboot.forceColdBoot: no
hw.cpu.ncore: 2
hw.keyboard: yes
hw.sensors.proximity: yes
hw.dPad: no
hw.lcd.height: 1920
vm.heapSize: 256
skin.dynamic: yes
hw.device.manufacturer: Google
hw.gps: yes
hw.audioInput: yes
image.sysdir.1: system-images/android-30/google_apis_playstore/x86/
showDeviceFrame: yes
hw.camera.back: virtualscene
AvdId: Pixel_2_API_30
hw.lcd.density: 420
hw.arc: false
hw.device.hash2: MD5:55acbc835978f326788ed66a5cd4c9a7
fastboot.forceChosenSnapshotBoot: no
fastboot.forceFastBoot: yes
hw.trackBall: no
hw.battery: yes
hw.sdCard: yes
tag.display: Google Play
runtime.network.latency: none
disk.dataPartition.size: 6442450944
hw.sensors.orientation: yes
avd.ini.encoding: UTF-8
hw.gpu.enabled: yes
```
</details>

<details>
<summary>my Android Java sample</summary>

```java
public class MainActivity extends AppCompatActivity {

    static private final int REQUEST_CODE_CHOOSE_IMAGE_FROM_GALLERY = 12;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        FloatingActionButton fab = findViewById(R.id.fab);
        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                final Intent pickImageIntent = new Intent(Intent.ACTION_GET_CONTENT);
                pickImageIntent.setType("image/*");

                if (check()) {
                    Log.e("main", "DID resolve");
                } else {
                    Log.e("main", "didn't resolve");
                }
                MainActivity.this.startActivityForResult(pickImageIntent, REQUEST_CODE_CHOOSE_IMAGE_FROM_GALLERY);
            }
        });
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        //noinspection SimplifiableIfStatement
        if (id == R.id.action_settings) {
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        switch (requestCode) {
            case REQUEST_CODE_CHOOSE_IMAGE_FROM_GALLERY:
                handleChooseImageResult(resultCode, data);
                break;
            default:
                return;
        }
    }

    private void handleChooseImageResult(int resultCode, Intent data) {
        if (resultCode == RESULT_OK && data != null) {
            Log.e("main", "ok");
            return;
        }
        Log.e("main", "KO");
    }

    private boolean check() {
        String[] allowedExtensions = null;
        final boolean isMultipleSelection = false;
        final Intent intent = new Intent(Intent.ACTION_GET_CONTENT);
        final Uri uri = Uri.parse(Environment.getExternalStorageDirectory().getPath() + File.separator);
        final String type = "image/*";
        intent.setDataAndType(uri, type);
        intent.setType(type);
        //intent.putExtra(Intent.EXTRA_ALLOW_MULTIPLE, this.isMultipleSelection);
        intent.addCategory(Intent.CATEGORY_OPENABLE);

        if (type.contains(",")) {
            allowedExtensions = type.split(",");
        }

        if (allowedExtensions != null) {
            intent.putExtra(Intent.EXTRA_MIME_TYPES, allowedExtensions);
        }

        return intent.resolveActivity(getPackageManager()) != null;
    }
}
```
</details>

<details>
<summary>build.gradle</summary>

```
apply plugin: 'com.android.application'

android {
    compileSdkVersion 30
    buildToolsVersion "30.0.2"

    defaultConfig {
        applicationId "com.example.myapplication"
        minSdkVersion 16
        targetSdkVersion 30
        versionCode 1
        versionName "1.0"

        testInstrumentationRunner "androidx.test.runner.AndroidJUnitRunner"
    }

    buildTypes {
        release {
            minifyEnabled false
            proguardFiles getDefaultProguardFile('proguard-android-optimize.txt'), 'proguard-rules.pro'
        }
    }
}

dependencies {
    implementation fileTree(dir: "libs", include: ["*.jar"])
    implementation 'androidx.appcompat:appcompat:1.1.0'
    implementation 'com.google.android.material:material:1.2.1'
    implementation 'androidx.constraintlayout:constraintlayout:1.1.3'
    implementation 'androidx.navigation:navigation-fragment:2.1.0'
    implementation 'androidx.navigation:navigation-ui:2.1.0'
    testImplementation 'junit:junit:4.12'
    androidTestImplementation 'androidx.test.ext:junit:1.1.2'
    androidTestImplementation 'androidx.test.espresso:espresso-core:3.2.0'

}
```
</details>

**Comment by miguelpruivo** [comment]
@monsieurtanuki do you have `targetSdk` defined in your build.gradle? If so, try removing (it's optional anyway) to see if it fixes. Just trying to understand what might be causing it.

**Comment by monsieurtanuki** [comment]
@miguelpruivo Correct assumption: if I remove `targetSdkVersion 30` from my `build.gradle`, it works (the `intent.resolveActivity()` returns not null).

**Comment by miguelpruivo** [comment]
This isn't new @monsieurtanuki, for some reason, setting `targetSdk` have resulted in some issues with different plugins, even with previous sdks. However, like I said, fortunately it's optional to include it. Nevertheless, I'll see what I can do.

Thanks!

**Comment by miguelpruivo** [comment]
@monsieurtanuki ok, I've just migrated to fully support SDK 30 nevertheless. I'll publish a new version 2.0.9 in a while with it.

Thank you for reporting and your detailed logs!

**Comment by miguelpruivo** [comment]
Fixed and published with [2.0.9](https://pub.dev/packages/file_picker/changelog).

Closing.

**Comment by bettkipronoh** [comment]
Am getting the same issue with version 2.1.6
